pkg/util/trivy/cache.go

```diff
@@ -9,16 +9,241 @@
+	gcTicker := time.NewTicker(garbageCollectionFrequency)
+	go func() {
+		for range gcTicker.C {
+			if err := db.RunValueLogGC(garbageCollectionDiscardRatio); err != nil && err != badger.ErrNoRewrite {
+				// ErrNoRewrite is returned when the GC runs fine but doesn't
+				// result any cleanup. That's why we don't log anything in that
+				// case.
+				log.Warnf("error while running garbage collection in Badger: %s", err)
+			}
+		}
+	}()
+
+	return &BadgerCache{db: db, ttl: ttl, garbageCollectionTicker: gcTicker}, nil
+}
+
+func (cache *BadgerCache) MissingBlobs(artifactID string, blobIDs []string) (bool, []string, error) {
+	var missingArtifact bool
+	var missingBlobIDs []string
+
+	err := cache.db.View(func(txn *badger.Txn) error {
+		if _, err := txn.Get([]byte(artifactID)); err != nil {
+			if err == badger.ErrKeyNotFound {
+				missingArtifact = true
+			} else {
+				return err
+			}
+		}
+
+		for _, blobID := range blobIDs {
+			if _, err := txn.Get([]byte(blobID)); err != nil {
+				if err == badger.ErrKeyNotFound {
+					missingBlobIDs = append(missingBlobIDs, blobID)
+				} else {
+					return err
+				}
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return false, nil, fmt.Errorf("error getting missing blobs from Badger cache: %w", err)
+	}
+
+	return missingArtifact, missingBlobIDs, nil
+}
+
+func (cache *BadgerCache) PutArtifact(artifactID string, artifactInfo types.ArtifactInfo) error {
+	artifactBytes, err := json.Marshal(artifactInfo)
+	if err != nil {
+		return fmt.Errorf("error converting artifact with ID %q to JSON: %w", artifactID, err)
+	}
+
+	err = cache.db.Update(func(txn *badger.Txn) error {
+		entry := badger.NewEntry([]byte(artifactID), artifactBytes).WithTTL(cache.ttl)
+		return txn.SetEntry(entry)
+	})
+	if err != nil {
+		return fmt.Errorf("error saving artifact with ID %q into Badger cache: %w", artifactID, err)
+	}
+
+	return nil
+}
+
+func (cache *BadgerCache) PutBlob(blobID string, blobInfo types.BlobInfo) error {
+	blobBytes, err := json.Marshal(blobInfo)
+	if err != nil {
+		return fmt.Errorf("error converting blob with ID %q to JSON: %w", blobID, err)
+	}
+
+	err = cache.db.Update(func(txn *badger.Txn) error {
+		entry := badger.NewEntry([]byte(blobID), blobBytes).WithTTL(cache.ttl)
+		return txn.SetEntry(entry)
+	})
+	if err != nil {
+		return fmt.Errorf("error saving blob with ID %q into Badger cache: %w", blobID, err)
+	}
+
+	return nil
+}
+
+func (cache *BadgerCache) DeleteBlobs(blobIDs []string) error {
+	var errs error
+
+	err := cache.db.Update(func(txn *badger.Txn) error {
+		for _, blobID := range blobIDs {
+			if err := txn.Delete([]byte(blobID)); err != nil {
+				errs = multierror.Append(errs, err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("error deleting blobs from Badger cache: %w", err)
+	}
+
+	return errs
+}
+
+func (cache *BadgerCache) GetArtifact(artifactID string) (types.ArtifactInfo, error) {
```

Comment:
`GetArtifact` and `GetBlob` seems to contain duplicated code.
May I suggest a way to factorize those functions in #15948 ?